### PR TITLE
ED-3071 get company by Companies House ID

### PR DESCRIPTION
**Pull request by kowalcj0**
This ticket [ED-3071](https://uktrade.atlassian.net/browse/ED-3071)

I've also added two test cases that use [hypothesis](https://hypothesis.readthedocs.io/en/latest/index.html) framework (kinda fuzzer) to test edge cases.

### CODE REVIEW

File: tests/test_testapi_client.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        response = self.client.get_company_by_ch_id(ch_id=None)
+        assert response is None
+
+    def test_should_return_none_on_empty_ch_id(self):
+        response = self.client.get_company_by_ch_id(ch_id="")
+        assert response is None
+
+    @mock.patch('directory_api_client.base.BaseAPIClient.request')
+    def test_should_not_make_any_request_on_empty_ch_id(self, mocked_request):
+        self.client.get_company_by_ch_id(ch_id="")
```
Comment (sdonk) [inline tests/test_testapi_client.py]: Please don’t mix `'` and `"`

File: directory_api_client/testapiclient.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+class DirectoryTestAPIClient(BaseAPIClient):
+
+    endpoints = {
+        'company_by_ch_id': 'testapi/company/{ch_id}/'
+    }
+
+    def get_company_by_ch_id(self, ch_id: str):
+        result = None
+        if ch_id:
```
Comment (richtier) [inline directory_api_client/testapiclient.py]: the name `get_company_by_ch_id` implies `ch_id` will always be present. Why allow `ch_id` to not be present here?

File: tests/test_testapi_client.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
+from unittest import TestCase, mock
+
+from tests import stub_request
+
+from directory_api_client.testapiclient import DirectoryTestAPIClient
+
+
+class DirectoryTestAPIClientTest(TestCase):
```
Comment (richtier) [inline tests/test_testapi_client.py]: the other test in this repo do not use `TestCase`. Please aim for consistency with the other tests.

Reasons not to use `TestCase` is it's like asking for a banana and being given a gorilla holding a banana. We can achieve what we want with simply doing:

```
@pytest.fixture
def client():
    base_url = 'https://example.com'
    key = 'test'
    client = DirectoryTestAPIClient(base_url, key)

def test_company(client):
    assert self.client.base_url == self.base_url
    assert self.client.request_signer.secret == self.key
```

the advantage is it's a lot flatter, and consistent with other tests.

File: directory_api_client/testapiclient.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
+from directory_api_client.base import BaseAPIClient
+
+
+class DirectoryTestAPIClient(BaseAPIClient):
+
+    endpoints = {
+        'company_by_ch_id': 'testapi/company/{ch_id}/'
+    }
+
+    def get_company_by_ch_id(self, ch_id: str):
```
Comment (richtier) [inline directory_api_client/testapiclient.py]: we don't use the type hinting anywhere else in API. If we're adding it, please create a refactoring ticket for adding it everywhere in API. https://uktrade.atlassian.net/browse/ED-3024

File: directory_api_client/testapiclient.py (hunk 2 above)
Comment (kowalcj0) [reply directory_api_client/testapiclient.py]: This check covers a case when you pass `ch_id=None`. It's a typical boundary value test https://en.wikipedia.org/wiki/Boundary_testing

File: tests/test_testapi_client.py (hunk 3 above)
Comment (kowalcj0) [reply tests/test_testapi_client.py]: @richtier all other `test_*` modules in this repo use `TestCase`. So, I'm actually consistent here.
I'd also prefer flatter structure, but that requires a bit of refactoring. Should we add a ticket for it?  https://uktrade.atlassian.net/browse/ED-3024

File: directory_api_client/testapiclient.py (hunk 4 above)
Comment (kowalcj0) [reply directory_api_client/testapiclient.py]: I've posted an answer to the same question here https://github.com/uktrade/directory-api/pull/386#discussion_r167511023

File: directory_api_client/testapiclient.py (hunk 2 above)
Comment (richtier) [reply directory_api_client/testapiclient.py]: ah I see. If we wanted to solve that problem it seems better to check the expected condition is met, and raising an error if not.

Returning `None` if the precondition is not met really is hiding the error. The error should be shouted loudly so we can fix it. Also, making this function sometimes return a `Response`, and sometimes return `None` is inconsistent behaviour. The fewer degrees of freedom the less we have to think :)

we could do this:

```
def get_company_by_ch_id(self, ch_id: str):
    assert ch_id
    ...
```

but still, I wonder if this is overkill. What is `ch_id` is dictionary or a list, or a boolean `True`? should we also be checking for that? No, I think not :)

I would remove the check that `ch_id` is `None`.

File: tests/test_testapi_client.py (hunk 3 above)
Comment (richtier) [reply tests/test_testapi_client.py]: ah oops. I've been spoilt testing in other repos :)

Comment (kowalcj0) [reply tests/test_testapi_client.py]: here's the ticket https://uktrade.atlassian.net/browse/ED-3394
